### PR TITLE
Change order of mplot3d view angles to match order of rotations

### LINKED ISSUE #28353: [MNT]: mplot3d azim-elev-roll order
### Summary

Part of PR #21426 'Add ability to roll the camera in 3D plots' (by Scott Shambaugh, 3 years ago) read as follows:
`- Make elev, azim, roll ordering consistent everywhere`

However, the choice of ordering is rather unfortunate:
- It is not the common one (most everyone else in the world uses azim-elev-roll, see below)
- It does not correspond to the order in which the 3 rotations take place (that's azim-elev-roll)
- It is not consistent with the ordering in matplotlib's `colors.py`.

I.e., there is room for improvement.

-----
 When I google:
- `"azimuth, elevation"` -> 291 000 results
- `"elevation, azimuth"`-> 41 500 results

That is to say, the `elevation, azimuth` ordering is not the common one. And if we consider that searching _for both_ yields:
- `"azimuth, elevation" and "elevation, azimuth"` -> 37 500 results (some links are non-committal), 

then it becomes clear that the `elevation, azimuth` ordering is really a minority position: (41-37)/(291-37) < 2%.
(Also, for what it's worth: MATLAB happens to use `"azimuth, elevation"`, just like (almost) everyone else.)

-----
The ordering with elevation first does not correspond to the order in which the 3 rotations take place:
if thinking about intrinsic rotations, then the rotation order is azim, elev, then roll.
The corresponding quaternion is
`    q = exp((+roll/2)*e_x)*exp((+elev/2)*e_y)*exp((-azim/2)*e_z)`
i.e., to be precise, mplot3d's `azim, elev, roll` are a kind of Tait-Bryan angles, -z,y',x".
It is a bit of a mess with the Euler angles, apparently there are 24 variations (not counting the signs) - https://github.com/moble/quaternion/wiki/Euler-angles-are-horrible

### Proposed fix

Switch (back) to the common ordering, the order in which the rotations are performed: `azim, elev, roll`.

Fortunately, Scott had the good foresight to use named parameters, so it is not a completely breaking change.
We can make the use of named parameters for `azim` and `elev` mandatory (preceding with a '*' argument), so changing the order (when parameters are unnamed) won't result in any silent failures.

Scott voiced a hesitation to change it (again), and I can understand his initial reaction. However, it seems to me that changing the order has clear benefits, and arguably it could be worth the effort of adapting existing code, as per https://matplotlib.org/devdocs/devel/api_changes.html (especially in view of the fact that the effort might be limited, with the named parameters).

I have a PR almost ready to show what the changes would be.

### PR DESCRIPTION
**Pull request by MischaMegens2**
## PR summary
Changes the order of the view angles to azim-elev-roll, since:
- azim-elev(-roll) appears to be a much more common order, the elev-azim(-roll) that we had is unusual 
- it corresponds to the order in which the 3 rotations take place
- it is consistent with the ordering in matplotlib's colors.py

The proposed changes (to the documentation of view_init() in particular) suggest to use keyword arguments 
in this order, but the historical positional arguments are still accepted without change, so the interface does not actually change with this PR.
Resolves #28353.

Summary of changes:
- Change order of mplot3d view angles to azim-elev-roll, the order in which rotations occur:
  - in axes3d.py
  - in tests
  - in documentation
  - in examples
- Add description of rotation to view_angles.rst
- Put in ref. to https://github.com/moble/quaternion/wiki/Euler-angles-are-horrible
- Update view_init() documentation
- Add doc\api\next_api_changes\deprecations warning that Axes3D.view_init preferably should be used with keyword arguments
- Add next_whats_new\view_angles_order.rst explaining the new view angles order

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [(please review)] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

File: doc/api/toolkits/mplot3d/view_angles.rst
Comment (scottshambaugh) [inline doc/api/toolkits/mplot3d/view_angles.rst:32]: I think the discussion around quaternions might be confusing, as no quaternion interface is exposed to the user. Perhaps something like the following? I haven't thought through the rotation handedness so that might need to be changed.

```
The azim, elev, roll rotation order corresponds to rotation of the axes in a fixed camera view frame. A left-hand rotation about camera z is applied (azim), then a right hand rotation about camera y (elev), then a right hand rotation about camera x (roll).

This can also be thought of orbiting a camera around fixed axes by flipping the order of operations. First the camera is rotated clockwise about the scene's +x axis (roll), then the +y axis (elev), then the -z axis (azim).
```

**Review by scottshambaugh — COMMENTED** [review]
Looks pretty good to me, one comment on the docs

File: lib/mpl_toolkits/mplot3d/axes3d.py
Comment (scottshambaugh) [inline lib/mpl_toolkits/mplot3d/axes3d.py:1158]: I like the structure of this change! Could you add a test for the positional args case?

File: doc/api/next_api_changes/deprecations/xxxxx-MM.rst
Comment (MischaMegens2) [reply doc/api/next_api_changes/deprecations/xxxxx-MM.rst:1]: Of course, and now that I have a PR number, it is straightforward to rename.

File: lib/mpl_toolkits/mplot3d/axes3d.py
Comment (MischaMegens2) [reply lib/mpl_toolkits/mplot3d/axes3d.py:1158]: A test has been added (test_axes3d.py::test_view_init_positional_args).

**Comment by oscargus** [comment]
An option may be to make these keyword only. This will introduce a smoother translation path, but take longer time. Once they are keyword only, the order doesn't matter... (So can be changed in the code base without any problems, nor benefits...)

File: doc/api/toolkits/mplot3d/view_angles.rst
Comment (MischaMegens2) [reply doc/api/toolkits/mplot3d/view_angles.rst:32]: I added something along the lines of your suggestion; the connection with quaternions is still mentioned but less prominently. Take a look.

**Comment by MischaMegens2** [comment]
Ready for review again.

**Review by scottshambaugh — APPROVED** [review]
Looks good to me now

**Comment by MischaMegens2** [comment]
Anything else needed for a merge? (Does that python 3.9 on macos-12 need to be revived?)

**Comment by MischaMegens2** [comment]
Added a test to improve coverage, and rebased. Ready for merge (I don't think the remaining failing checks have anything to do with the PR).

**Comment by MischaMegens2** [comment]
ping @scottshambaugh Could you take a look at it, or ask someone else for additional review (timhoffm?)

**Comment by scottshambaugh** [comment]
Hey @MischaMegens2, it's got my approval. For MRs we require approvals from 2 maintainers before it can be merged in. It can be a little frustrating to have good-to-go MRs sitting waiting for a bit unlooked at, but since everyone here is volunteering time that's for better or worse the nature of the beast, and it can take some time for someone to get around to it. It doesn't really help that the main branch CI is failing right now so it's not so easy to scan the MRs and see where has everything passing.

If you want to prod it along, especially if an MR is blocking a future improvement you want to work on, there are basically 3 avenues to do so:
1) Request a review from someone in the top right menu, or @ them directly
2) Ask for a review on the project chat on gitter: https://app.gitter.im/#/room/#matplotlib_matplotlib:gitter.im
3) Hop on the weekly zoom meeting and bring it up: https://hackmd.io/l9vkn_T4RSmk147H_ZPPBA

In this case since @timhoffm was active on the original issue, he's probably a good one to ping for a review here.

**Comment by MischaMegens2** [comment]
Ping @timhoffm  Dear Tim, would you be able to take a look and provide a second review?

**Comment by timhoffm** [comment]
I'm not convinced of using a different order for keyword arguments than their positional definition. It's a common expectation that parameters are listed in the order they are defined in the signature. When seeing the following both notations, it's not uncommon  to assume that -35 in the second example will be `azim`.
```
ax.view_init(azim=-35, elev=20)
ax.view_init(-35, 20)
```
Since users will see both notations, or try to shorten from the first to the second to save typing, I believe we're doing more harm than good by reordering the official docs.


I'm sorry, I have not read/thought carefully enough [here](https://github.com/matplotlib/matplotlib/issues/28353#issuecomment-2153423466)

> > But we could nudge in the right direction - update the examples to use the keyword arguments, encourage their use in the documentation (and use the azim, elev order in the mplot3d code itself).
>
> Agreed. This is an improvement anyway. Do you want to make a PR?

I believe we should use kwargs throughout to make the "unconventional order" transparent. But we should not try to gloss over it by reordering the arguments when calling. Since we cannot change the order in the forseeable future, I'd rather have users slightly surprised by seeing the "unconventional order", rather than making them believe everything is conventional and then stumble over positional order. That's a didactic aspect of our docs. Of course, users are free to use any order they like in their code.

**Comment by MischaMegens2** [comment]
@timhoffm: Do I understand correctly then that
- the `view_init()` suggestion is fine in your opinion, but
- the changing of the order in the examples you think is a bridge too far at this point

(The change to the examples should come later?)
 
> I'm not convinced of using a different order for keyword arguments than their positional definition. It's a common expectation that parameters are listed in the order they are defined in the signature. When seeing the following both notations, it's not uncommon to assume that -35 in the second example will be `azim`.
> 
> ```
> ax.view_init(azim=-35, elev=20)
> ax.view_init(-35, 20)
> ```
> 

Well yes, it would be an assumption, and an unwarranted assumption - as you said, people are free to use any order they like for keyword arguments...

> Since users will see both notations, or try to shorten from the first to the second to save typing, [...]

Such shortening would be ill-advised of course; but I don't quite understand why that would be desirable, what's the appeal. 'There is no tax on keystrokes' (Bertrand Meyer).
Fortunately, if anyone did so inadvertently, then they might stumble momentarily but wouldn't fall, they would be saved by:
- Add doc\api\next_api_changes\deprecations warning that Axes3D.view_init preferably should be used with keyword arguments.

So I thought we are at liberty to do the two steps above simultaneously, without introducing grave risks for mistakes.

**Comment by timhoffm** [comment]
Oh, no, communication has gone quite wrong here. My appologies, I have not been reading and writing carefully enough (too much going on and trying to quickly respond in between 😢).

My standpoint is that there is so much current positional use that we cannot afford to bother existing users with changes. With any such change, we would impose work on current users. That also includes a runtime warning for positional use. While not breaking, users would still have to modify their code, because you don't want to have code with warnings.

The *only* thing we should do, is change our internal examples to use kwargs, to make the users aware of the meaning and order of the parameters. Here, I would still use the order as in the signature to avoid implicit surprises: `ax.view_init(azim=-35, elev=20)`.

If really wanted we could add a note to the docstring:

> It's recommended to provide the parameter as keyword arguments ` `.view_init(elev=20, azim=-35)``. The positional order is unfortunately different from that used in other programs ``(azim, elev)``, but cannot be changed du to backward compatibility.

But I'm 50/50 on that.

Unfortunately, this also means there is no migration path and we'll have to live with the unconventional order. It's a tradeoff between a suboptimal API and backward compatibility; for this case, I'm clearly on the side of not annoying existing users. You would have to convince other core developers that this change is worth the churn.

**Review by timhoffm — CHANGES_REQUESTED** [review]
I block this based on https://github.com/matplotlib/matplotlib/pull/28395#issuecomment-2220000516

If others think this is an acceptable change, please review.

**Comment by MischaMegens2** [comment]
> The _only_ thing we should do, is change our internal examples to use kwargs, to make the users aware of the meaning and order of the parameters. Here, I would still use the order as in the signature to avoid implicit surprises: `ax.view_init(azim=-35, elev=20)`.

Ah, I can probably come up with a less disruptive approach, shortly.

**Comment by MischaMegens2** [comment]
(The remaining CircleCI warnings are:
```
/home/circleci/project/lib/mpl_toolkits/mplot3d/axes3d.py:docstring of mpl_toolkits.mplot3d.axes3d.Axes3D.view_init:23: WARNING: py:obj reference target not found: azim, elev
/home/circleci/project/doc/users/next_whats_new/view_angles_keyword_arguments.rst:4: WARNING: py:obj reference target not found: elev, azim
```
Not quite sure what to do about those, please advise...)

**Comment by timhoffm** [comment]
> (The remaining CircleCI warnings are:
> 
> ```
> /home/circleci/project/lib/mpl_toolkits/mplot3d/axes3d.py:docstring of mpl_toolkits.mplot3d.axes3d.Axes3D.view_init:23: WARNING: py:obj reference target not found: azim, elev
> /home/circleci/project/doc/users/next_whats_new/view_angles_keyword_arguments.rst:4: WARNING: py:obj reference target not found: elev, azim
> ```
> 
> Not quite sure what to do about those, please advise...)

You need to use double backticks for literal quoting. Single backticks try to link to code objects, which don't exist and thus warn.

File: galleries/examples/mplot3d/rotate_axes3d_sgskip.py
Comment (timhoffm) [inline galleries/examples/mplot3d/rotate_axes3d_sgskip.py:47]: +/-0 on using kwargs here. The use of kwargs is meaningful with pure numbers to communicate meaning and order of the parameters, but we already get that through the variable names.

File: doc/users/next_whats_new/view_angles_keyword_arguments.rst
Comment (timhoffm) [inline doc/users/next_whats_new/view_angles_keyword_arguments.rst:7]: This is outdated. We do not want to change the order. I believe the whole waht's new note should go. Instead put a note on the unconventional order in view_angles.rst and/or the view_init docstring.

**Review by timhoffm — COMMENTED** [review]
Global comment: we have a fundamental discrepancy in angle order between our API (which as discussed we cannot change) and the common convention "out there". For each usage we have to decide, which one we want to be consistent with. I favor internal consistency. In degrees of emphasis

-1 on order changes in public API
-0.5 on order changes non-public code
-0.1 on order changes in non-code changes (e.g. written documentation)

File: lib/mpl_toolkits/mplot3d/tests/test_axes3d.py
Comment (MischaMegens2) [reply lib/mpl_toolkits/mplot3d/tests/test_axes3d.py:135]: Makes sense, will do.

File: galleries/examples/mplot3d/rotate_axes3d_sgskip.py
Comment (MischaMegens2) [reply galleries/examples/mplot3d/rotate_axes3d_sgskip.py:47]: It seems just repetitive now, and not longer serves a purpose, I'll get rid of it, change it back to the original (also in other places).

File: lib/mpl_toolkits/mplot3d/axes3d.py
Comment (MischaMegens2) [reply lib/mpl_toolkits/mplot3d/axes3d.py:1588]: ...well actually, the order _is_ the order that the _Quaternion API defines, and the change is as intended. Since the quaternion class has its own API, it is not bound by my previous mistake to force it to be consistent with the Axes3D class. And furthermore, _Quaternion is an internal class, so I have little hesitation to change it back to the way it was initially, with a logical progression of arguments for `from_cardan_angles()` and `as_cardan_angles()`.

File: lib/mpl_toolkits/mplot3d/axes3d.py
Comment (MischaMegens2) [reply lib/mpl_toolkits/mplot3d/axes3d.py:1597]: See above.

File: doc/api/toolkits/mplot3d/view_angles.rst
Comment (MischaMegens2) [reply doc/api/toolkits/mplot3d/view_angles.rst:44]: I considered it, but I get the impression that  `..note::` is not meant for the purpose of adding some reference material which is a bit secondary, as it seems is our intention; rather, it is for [information you want the user to pay particular attention to](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html). I wouldn't want to put that much emphasis on the particulars of the definition of the view angles.

Since the text already mentions that '[The view angles] are further documented in the `.mplot3d.axes3d.Axes3D.view_init` API', I thought it be sensible to move the information there (and keep the view_angles.rst as it was). I'll give it a try.

File: doc/api/toolkits/mplot3d/view_angles.rst
Comment (MischaMegens2) [reply doc/api/toolkits/mplot3d/view_angles.rst:31]: I followed your suggestion below; the extra section is gone from view_angles.rst

File: doc/users/next_whats_new/view_angles_keyword_arguments.rst
Comment (MischaMegens2) [reply doc/users/next_whats_new/view_angles_keyword_arguments.rst:7]: Removed it.

**Comment by MischaMegens2** [comment]
@timhoffm: Should be good to go, please take a look. (I left the 'Resolve conversation' up to you.)

File: galleries/examples/mplot3d/view_planes_3d.py
Comment (scottshambaugh) [inline galleries/examples/mplot3d/view_planes_3d.py:19]: This change removes the ability to copy-paste these docs into ones functions, and is confusing wrt the keyword ordering below.

**Review by scottshambaugh — COMMENTED** [review]
After looking at this again, I think the changes here make things more confusing than not. There are so many Euler/cardan angle orderings for 3D orientations out there, that confirming a specific convention before use is expected. IMO the pedagogical burden is far less for users by giving them complete internal consistency rather than partially aligning with rotation ordering used elsewhere. I think action to make things clearer should focus on improving the docs and examples, without changing the keyword argument ordering.

Greatly appreciate the effort and thought that went into this PR, but I'm a -1 on merging it in.

**Comment by timhoffm** [comment]
I argee with @scottshambaugh. Let's prioritize internal consistency within Matplotlib.

**Comment by MischaMegens2** [comment]
> I argee with @scottshambaugh. Let's prioritize internal consistency within Matplotlib.

How about: I revise it to just clarifying the situation in the documentation?

**Comment by ksunden** [comment]
Pushed tentatively to 3.11 as we are trying to close out 3.10, though if (as it sounds like) this turns into doc-only changes, then certainly willing to put into the 3.10.x-doc branch.

**Comment by MischaMegens2** [comment]
> Pushed tentatively to 3.11 as we are trying to close out 3.10, though if (as it sounds like) this turns into doc-only changes, then certainly willing to put into the 3.10.x-doc branch.

Yes, doc-only changes. What is the time frame for the 3.10.x-doc ?

**Comment by ksunden** [comment]
Doc changes are merged when they are ready, don't necessarily follow the release schedule. I intend to branch soon, but essentially doc only means it can be backported even without a patch release being needed. It's mostly a distinction for maintainers. If this PR were to be included in its current state, which changes parameter orders, etc. It would likely land in the next `.0` (meso) release.

**Comment by timhoffm** [comment]
> essentially doc only means it can be backported even without a patch release being needed.

If this contains docstring changes on anything in `lib`, it doesn't qualify for the doc branch and would be targeted for the next micro release 3.10.x.

https://matplotlib.org/devdocs/devel/pr_guide.html#backport-strategy

**Comment by MischaMegens2** [comment]
> If this contains docstring changes on anything in `lib`, it doesn't qualify for the doc branch and would be targeted for the next micro release 3.10.x.

Oh, it would contain docstring changes in `lib` (of course), good to know!

**Comment by MischaMegens2** [comment]
There is a new PR #29114 to replace this one, let's close this one.

**Pull request closed without merging**
